core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2022,29 +2022,84 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    def arbiter(
+      promise: Promise[Nothing, Either[Cause[E], List[B]]],
+      buffer: Ref[Array[B]],
+      todo: Ref[Int],
+      idx: Int
+    )(res: Exit[E, B]): ZIO[R, Nothing, Unit] =
+      res.foldM[R, Nothing, Unit](
+        e => promise.succeed(Left(e)).unit,
+        a =>
+          todo.modify { t =>
+            val updateBuffer = buffer.modify(b => (b.update(idx, a), b))
+            if (t == 1) (updateBuffer <* buffer.get.flatMap(b => promise.succeed(Right(b.toList)))) -> 0
+            else updateBuffer                                                                       -> (t - 1)
+          }.flatten
+      )
+
+    (for {
+      size    <- UIO.effectTotal(as.size)
+      todo    <- Ref.make(size)
+      buffer  <- Ref.make(new Array[Any](size).asInstanceOf[Array[B]])
+      promise <- Promise.make[Nothing, Either[Cause[E], List[B]]]
+      c <- ZIO.uninterruptibleMask { restore =>
+            for {
+              as <- ZIO.traverse(as.zipWithIndex) {
+                     case (a, idx) =>
+                       ZIO
+                         .interruptible(fn(a))
+                         .run
+                         .flatMap { exit =>
+                           arbiter(promise, buffer, todo, idx)(exit)
+                         }
+                         .fork
+                   }
+              _ <- promise.succeed(Right(Nil)).when(as.isEmpty)
+              c <- restore(promise.await.absolve.unsandbox)
+                    .onError(_ => promise.interrupt *> Fiber.interruptAll(as))
```

Comment:
I think intention is to uninterruptibly fork fibers and register `aribter` on their completion but `await`ing and later can be interrupted.